File: app/src/main/java/be/digitalia/fosdem/db/ScheduleDao.java
package be.digitalia.fosdem.db;

import android.app.SearchManager;
import android.database.Cursor;
import android.provider.BaseColumns;

import androidx.annotation.MainThread;
import androidx.annotation.Nullable;
import androidx.annotation.WorkerThread;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.paging.DataSource;
import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Transaction;

import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import be.digitalia.fosdem.alarms.FosdemAlarmManager;
import be.digitalia.fosdem.db.entities.EventEntity;
import be.digitalia.fosdem.db.entities.EventTitles;
import be.digitalia.fosdem.db.entities.EventToPerson;
import be.digitalia.fosdem.model.Day;
import be.digitalia.fosdem.model.DetailedEvent;
import be.digitalia.fosdem.model.Event;
import be.digitalia.fosdem.model.EventDetails;
import be.digitalia.fosdem.model.Link;
import be.digitalia.fosdem.model.Person;
import be.digitalia.fosdem.model.StatusEvent;
import be.digitalia.fosdem.model.Track;
import be.digitalia.fosdem.utils.DateUtils;

@Dao
public abstract class ScheduleDao {

	private static final String LAST_UPDATE_TIME_PREF = "last_update_time";
	private static final String LAST_MODIFIED_TAG_PREF = "last_modified_tag";

	private final AppDatabase appDatabase;
	private final MutableLiveData<Long> lastUpdateTime = new MutableLiveData<>();

	public ScheduleDao(AppDatabase appDatabase) {
		this.appDatabase = appDatabase;
	}

	/**
	 * @return The last update time in milliseconds since EPOCH, or -1 if not available.
	 * This LiveData is pre-initialized with the up-to-date value.
	 */
	@MainThread
	public LiveData<Long> getLastUpdateTime() {
		if (lastUpdateTime.getValue() == null) {
			lastUpdateTime.setValue(appDatabase.getSharedPreferences().getLong(LAST_UPDATE_TIME_PREF, -1L));
		}
		return lastUpdateTime;
	}

	/**
	 * @return The time identifier of the current version of the database.
	 */
	public String getLastModifiedTag() {
		return appDatabase.getSharedPreferences().getString(LAST_MODIFIED_TAG_PREF, null);
	}

	private static class EmptyScheduleException extends RuntimeException {
	}

	/**
	 * Stores the schedule in the database.
	 *
	 * @param events The events stream.
	 * @return The number of events processed.
	 */
	@WorkerThread
	public int storeSchedule(Iterable<DetailedEvent> events, String lastModifiedTag) {
		int totalEvents;
		try {
			totalEvents = storeScheduleInternal(events, lastModifiedTag);
		} catch (EmptyScheduleException ese) {
			totalEvents = 0;
		}
		if (totalEvents > 0) {
			// Set last update time and server's last modified tag
			final long now = System.currentTimeMillis();
			appDatabase.getSharedPreferences().edit()
					.putLong(LAST_UPDATE_TIME_PREF, now)
					.putString(LAST_MODIFIED_TAG_PREF, lastModifiedTag)
					.apply();
			lastUpdateTime.postValue(now);

			FosdemAlarmManager.INSTANCE.onScheduleRefreshed();
		}
		return totalEvents;
	}

	@Transaction
	protected int storeScheduleInternal(Iterable<DetailedEvent> events, String lastModifiedTag) {
		// 1: Delete the previous schedule
		clearSchedule();

		// 2: Insert the events
		int totalEvents = 0;
		final Map<Track, Long> tracks = new HashMap<>();
		long nextTrackId = 0L;
		long minEventId = Long.MAX_VALUE;
		final Set<Day> days = new HashSet<>(2);

		for (DetailedEvent detailedEvent : events) {
			final Event event = detailedEvent.getEvent();
			final EventDetails details = detailedEvent.getDetails();
			// Retrieve or insert Track
			final Track track = event.getTrack();
			Long trackId = tracks.get(track);
			if (trackId == null) {
				// New track
				nextTrackId++;
				trackId = nextTrackId;
				final Track newTrack = new Track(nextTrackId, track.getName(), track.getType());
				insertTrack(newTrack);
				tracks.put(newTrack, trackId);
			}

			final long eventId = event.getId();
			try {
				// Insert main event and fulltext fields
				final EventEntity eventEntity = new EventEntity(
						eventId,
						event.getDay().getIndex(),
						event.getStartTime(),
						event.getEndTime(),
						event.getRoomName(),
						event.getSlug(),
						trackId,
						event.getAbstractText(),
						event.getDescription()
				);
				final EventTitles eventTitles = new EventTitles(
						eventId,
						event.getTitle(),
						event.getSubTitle()
				);
				insertEvent(eventEntity, eventTitles);
			} catch (Exception e) {
				// Duplicate event: skip
				continue;
			}

			days.add(event.getDay());
			if (eventId < minEventId) {
				minEventId = eventId;
			}

			final List<Person> persons = details.getPersons();
			insertPersons(persons);
			final int personsCount = persons.size();
			final EventToPerson[] eventsToPersons = new EventToPerson[personsCount];
			for (int i = 0; i < personsCount; ++i) {
				eventsToPersons[i] = new EventToPerson(eventId, persons.get(i).getId());
			}
			insertEventsToPersons(eventsToPersons);

			insertLinks(details.getLinks());

			totalEvents++;
		}

		if (totalEvents == 0) {
			// Rollback the transaction
			throw new EmptyScheduleException();
		}

		// 3: Insert collected days
		insertDays(days);

		// 4: Purge outdated bookmarks
		purgeOutdatedBookmarks(minEventId);

		return totalEvents;
	}

	@Insert
	protected abstract void insertTrack(Track track);

	@Insert
	protected abstract void insertEvent(EventEntity eventEntity, EventTitles eventTitles);

	@Insert(onConflict = OnConflictStrategy.IGNORE)
	protected abstract void insertPersons(List<Person> persons);

	@Insert
	protected abstract void insertEventsToPersons(EventToPerson[] eventsToPersons);

	@Insert
	protected abstract void insertLinks(List<Link> links);

	@Insert
	protected abstract void insertDays(Set<Day> days);

	@Query("DELETE FROM bookmarks WHERE event_id < :minEventId")
	protected abstract void purgeOutdatedBookmarks(long minEventId);

	@WorkerThread
	@Transaction
	public void clearSchedule() {
		clearEvents();
		clearEventTitles();
		clearPersons();
		clearEventToPersons();
		clearLinks();
		clearTracks();
		clearDays();
	}

	@Query("DELETE FROM events")
	protected abstract void clearEvents();

	@Query("DELETE FROM events_titles")
	protected abstract void clearEventTitles();

	@Query("DELETE FROM persons")
	protected abstract void clearPersons();

	@Query("DELETE FROM events_persons")
	protected abstract void clearEventToPersons();

	@Query("DELETE FROM links")
	protected abstract void clearLinks();

	@Query("DELETE FROM tracks")
	protected abstract void clearTracks();

	@Query("DELETE FROM days")
	protected abstract void clearDays();


	// Cache days
	private volatile LiveData<List<Day>> daysLiveData;

	public LiveData<List<Day>> getDays() {
		if (daysLiveData != null) {
			return daysLiveData;
		}
		synchronized (this) {
			daysLiveData = getDaysInternal();
			return daysLiveData;
		}
	}

	@Query("SELECT `index`, date FROM days ORDER BY `index` ASC")
	protected abstract LiveData<List<Day>> getDaysInternal();

	@WorkerThread
	public int getYear() {
		long date = 0L;

		// Compute from cached days if available
		final LiveData<List<Day>> cache = daysLiveData;
		List<Day> days = (cache == null) ? null : cache.getValue();
		if (days != null) {
			if (days.size() > 0) {
				date = days.get(0).getDate().getTime();
			}
		} else {
			date = getConferenceStartDate();
		}

		// Use the current year by default
		if (date == 0L) {
			date = System.currentTimeMillis();
		}

		return DateUtils.INSTANCE.getYear(date);
	}

	@Query("SELECT date FROM days ORDER BY `index` ASC LIMIT 1")
	protected abstract long getConferenceStartDate();

	@Query("SELECT t.id, t.name, t.type FROM tracks t"
			+ " JOIN events e ON t.id = e.track_id"
			+ " WHERE e.day_index = :day"
			+ " GROUP BY t.id"
			+ " ORDER BY t.name ASC")
	public abstract LiveData<List<Track>> getTracks(Day day);

	/**
	 * Returns the event with the specified id, or null if not found.
	 */
	@Query("SELECT e.id, e.start_time, e.end_time, e.room_name, e.slug, et.title, et.subtitle, e.abstract, e.description"
			+ ", GROUP_CONCAT(p.name, ', ') AS persons, e.day_index, d.date AS day_date, e.track_id, t.name AS track_name, t.type AS track_type"
			+ " FROM events e"
			+ " JOIN events_titles et ON e.id = et.`rowid`"
			+ " JOIN days d ON e.day_index = d.`index`"
			+ " JOIN tracks t ON e.track_id = t.id"
			+ " LEFT JOIN events_persons ep ON e.id = ep.event_id"
			+ " LEFT JOIN persons p ON ep.person_id = p.`rowid`"
			+ " WHERE e.id = :id"
			+ " GROUP BY e.id")
	@Nullable
	@WorkerThread
	public abstract Event getEvent(long id);

	/**
	 * Returns all found events whose id is part of the given list.
	 */
	@Query("SELECT e.id, e.start_time, e.end_time, e.room_name, e.slug, et.title, et.subtitle, e.abstract, e.description"
			+ ", GROUP_CONCAT(p.name, ', ') AS persons, e.day_index, d.date AS day_date, e.track_id, t.name AS track_name, t.type AS track_type"
			+ ", b.event_id IS NOT NULL AS is_bookmarked"
			+ " FROM events e"
			+ " JOIN events_titles et ON e.id = et.`rowid`"
			+ " JOIN days d ON e.day_index = d.`index`"
			+ " JOIN tracks t ON e.track_id = t.id"
			+ " LEFT JOIN events_persons ep ON e.id = ep.event_id"
			+ " LEFT JOIN persons p ON ep.person_id = p.`rowid`"
			+ " LEFT JOIN bookmarks b ON e.id = b.event_id"
			+ " WHERE e.id IN (:ids)"
			+ " GROUP BY e.id"
			+ " ORDER BY e.start_time ASC")
	public abstract DataSource.Factory<Integer, StatusEvent> getEvents(long[] ids);

	/**
	 * Returns the events for a specified track.
	 */
	@Query("SELECT e.id, e.start_time, e.end_time, e.room_name, e.slug, et.title, et.subtitle, e.abstract, e.description"
			+ ", GROUP_CONCAT(p.name, ', ') AS persons, e.day_index, d.date AS day_date, e.track_id, t.name AS track_name, t.type AS track_type"
			+ ", b.event_id IS NOT NULL AS is_bookmarked"
			+ " FROM events e"
			+ " JOIN events_titles et ON e.id = et.`rowid`"
			+ " JOIN days d ON e.day_index = d.`index`"
			+ " JOIN tracks t ON e.track_id = t.id"
			+ " LEFT JOIN events_persons ep ON e.id = ep.event_id"
			+ " LEFT JOIN persons p ON ep.person_id = p.`rowid`"
			+ " LEFT JOIN bookmarks b ON e.id = b.event_id"
			+ " WHERE e.day_index = :day AND e.track_id = :track"
			+ " GROUP BY e.id"
			+ " ORDER BY e.start_time ASC")
	public abstract LiveData<List<StatusEvent>> getEvents(Day day, Track track);

	/**
	 * Returns a snapshot of the events for a specified track (without the bookmark status).
	 */
	@Query("SELECT e.id, e.start_time, e.end_time, e.room_name, e.slug, et.title, et.subtitle, e.abstract, e.description"
			+ ", GROUP_CONCAT(p.name, ', ') AS persons, e.day_index, d.date AS day_date, e.track_id, t.name AS track_name, t.type AS track_type"
			+ " FROM events e"
			+ " JOIN events_titles et ON e.id = et.`rowid`"
			+ " JOIN days d ON e.day_index = d.`index`"
			+ " JOIN tracks t ON e.track_id = t.id"
			+ " LEFT JOIN events_persons ep ON e.id = ep.event_id"
			+ " LEFT JOIN persons p ON ep.person_id = p.`rowid`"
			+ " WHERE e.day_index = :day AND e.track_id = :track"
			+ " GROUP BY e.id"
			+ " ORDER BY e.start_time ASC")
	public abstract List<Event> getEventsSnapshot(Day day, Track track);

	/**
	 * Returns events starting in the specified interval, ordered by ascending start time.
	 */
	@Query("SELECT e.id, e.start_time, e.end_time, e.room_name, e.slug, et.title, et.subtitle, e.abstract, e.description"
			+ ", GROUP_CONCAT(p.name, ', ') AS persons, e.day_index, d.date AS day_date, e.track_id, t.name AS track_name, t.type AS track_type"
			+ ", b.event_id IS NOT NULL AS is_bookmarked"
			+ " FROM events e"
			+ " JOIN events_titles et ON e.id = et.`rowid`"
			+ " JOIN days d ON e.day_index = d.`index`"
			+ " JOIN tracks t ON e.track_id = t.id"
			+ " LEFT JOIN events_persons ep ON e.id = ep.event_id"
			+ " LEFT JOIN persons p ON ep.person_id = p.`rowid`"
			+ " LEFT JOIN bookmarks b ON e.id = b.event_id"
			+ " WHERE e.start_time BETWEEN :minStartTime AND :maxStartTime"
			+ " GROUP BY e.id"
			+ " ORDER BY e.start_time ASC")
	public abstract DataSource.Factory<Integer, StatusEvent> getEventsWithStartTime(long minStartTime, long maxStartTime);

	/**
	 * Returns events in progress at the specified time, ordered by descending start time.
	 */
	@Query("SELECT e.id, e.start_time, e.end_time, e.room_name, e.slug, et.title, et.subtitle, e.abstract, e.description"
			+ ", GROUP_CONCAT(p.name, ', ') AS persons, e.day_index, d.date AS day_date, e.track_id, t.name AS track_name, t.type AS track_type"
			+ ", b.event_id IS NOT NULL AS is_bookmarked"
			+ " FROM events e"
			+ " JOIN events_titles et ON e.id = et.`rowid`"
			+ " JOIN days d ON e.day_index = d.`index`"
			+ " JOIN tracks t ON e.track_id = t.id"
			+ " LEFT JOIN events_persons ep ON e.id = ep.event_id"
			+ " LEFT JOIN persons p ON ep.person_id = p.`rowid`"
			+ " LEFT JOIN bookmarks b ON e.id = b.event_id"
			+ " WHERE e.start_time <= :time AND :time < e.end_time"
			+ " GROUP BY e.id"
			+ " ORDER BY e.start_time DESC")
	public abstract DataSource.Factory<Integer, StatusEvent> getEventsInProgress(long time);

	/**
	 * Returns the events presented by the specified person.
	 */
	@Query("SELECT e.id , e.start_time, e.end_time, e.room_name, e.slug, et.title, et.subtitle, e.abstract, e.description"
			+ ", GROUP_CONCAT(p.name, ', ') AS persons, e.day_index, d.date AS day_date, e.track_id, t.name AS track_name, t.type AS track_type"
			+ ", b.event_id IS NOT NULL AS is_bookmarked"
			+ " FROM events e"
			+ " JOIN events_titles et ON e.id = et.`rowid`"
			+ " JOIN days d ON e.day_index = d.`index`"
			+ " JOIN tracks t ON e.track_id = t.id"
			+ " LEFT JOIN events_persons ep ON e.id = ep.event_id"
			+ " LEFT JOIN persons p ON ep.person_id = p.`rowid`"
			+ " LEFT JOIN bookmarks b ON e.id = b.event_id"
			+ " JOIN events_persons ep2 ON e.id = ep2.event_id"
			+ " WHERE ep2.person_id = :person"
			+ " GROUP BY e.id"
			+ " ORDER BY e.start_time ASC")
	public abstract DataSource.Factory<Integer, StatusEvent> getEvents(Person person);

	/**
	 * Search through matching titles, subtitles, track names, person names. We need to use an union of 3 sub-queries because a "match" condition can not be
	 * accompanied by other conditions in a "where" statement.
	 */
	@Query("SELECT e.id, e.start_time, e.end_time, e.room_name, e.slug, et.title, et.subtitle, e.abstract, e.description"
			+ ", GROUP_CONCAT(p.name, ', ') AS persons, e.day_index, d.date AS day_date, e.track_id, t.name AS track_name, t.type AS track_type"
			+ ", b.event_id IS NOT NULL AS is_bookmarked"
			+ " FROM events e"
			+ " JOIN events_titles et ON e.id = et.`rowid`"
			+ " JOIN days d ON e.day_index = d.`index`"
			+ " JOIN tracks t ON e.track_id = t.id"
			+ " LEFT JOIN events_persons ep ON e.id = ep.event_id"
			+ " LEFT JOIN persons p ON ep.person_id = p.`rowid`"
			+ " LEFT JOIN bookmarks b ON e.id = b.event_id"
			+ " WHERE e.id IN ( "
			+ "SELECT `rowid`"
			+ " FROM events_titles"
			+ " WHERE events_titles MATCH :query || '*'"
			+ " UNION "
			+ "SELECT e.id"
			+ " FROM events e"
			+ " JOIN tracks t ON e.track_id = t.id"
			+ " WHERE t.name LIKE '%' || :query || '%'"
			+ " UNION "
			+ "SELECT ep.event_id"
			+ " FROM events_persons ep"
			+ " JOIN persons p ON ep.person_id = p.`rowid`"
			+ " WHERE p.name MATCH :query || '*'"
			+ " )"
			+ " GROUP BY e.id"
			+ " ORDER BY e.start_time ASC")
	public abstract DataSource.Factory<Integer, StatusEvent> getSearchResults(String query);

	/**
	 * Method called by SearchSuggestionProvider to return search results in the format expected by the search framework.
	 */
	@Query("SELECT e.id AS " + BaseColumns._ID
			+ ", et.title AS " + SearchManager.SUGGEST_COLUMN_TEXT_1
			+ ", IFNULL(GROUP_CONCAT(p.name, ', '), '') || ' - ' || t.name AS " + SearchManager.SUGGEST_COLUMN_TEXT_2
			+ ", e.id AS " + SearchManager.SUGGEST_COLUMN_INTENT_DATA
			+ " FROM events e"
			+ " JOIN events_titles et ON e.id = et.`rowid`"
			+ " JOIN tracks t ON e.track_id = t.id"
			+ " LEFT JOIN events_persons ep ON e.id = ep.event_id"
			+ " LEFT JOIN persons p ON ep.person_id = p.`rowid`"
			+ " WHERE e.id IN ( "
			+ "SELECT `rowid`"
			+ " FROM events_titles"
			+ " WHERE events_titles MATCH :query || '*'"
			+ " UNION "
			+ "SELECT e.id"
			+ " FROM events e"
			+ " JOIN tracks t ON e.track_id = t.id"
			+ " WHERE t.name LIKE '%' || :query || '%'"
			+ " UNION "
			+ "SELECT ep.event_id"
			+ " FROM events_persons ep"
			+ " JOIN persons p ON ep.person_id = p.`rowid`"
			+ " WHERE p.name MATCH :query || '*'"
			+ " )"
			+ " GROUP BY e.id"
			+ " ORDER BY e.start_time ASC LIMIT :limit")
	@WorkerThread
	public abstract Cursor getSearchSuggestionResults(String query, int limit);

	/**
	 * Returns all persons in alphabetical order.
	 */
	@Query("SELECT `rowid`, name"
			+ " FROM persons"
			+ " ORDER BY name COLLATE NOCASE")
	public abstract DataSource.Factory<Integer, Person> getPersons();

	public LiveData<EventDetails> getEventDetails(final Event event) {
		final MutableLiveData<EventDetails> result = new MutableLiveData<>();
		appDatabase.getQueryExecutor().execute(() -> result.postValue(new EventDetails(getPersons(event), getLinks(event))));
		return result;
	}

	@Query("SELECT p.`rowid`, p.name"
			+ " FROM persons p"
			+ " JOIN events_persons ep ON p.`rowid` = ep.person_id"
			+ " WHERE ep.event_id = :event")
	protected abstract List<Person> getPersons(Event event);

	@Query("SELECT * FROM links WHERE event_id = :event ORDER BY id ASC")
	protected abstract List<Link> getLinks(Event event);
}

File: app/src/main/java/be/digitalia/fosdem/db/ScheduleDao.kt
package be.digitalia.fosdem.db

import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.longPreferencesKey
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.paging.PagingSource
import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import androidx.room.Transactor.SQLiteTransactionType
import androidx.room.TypeConverters
import androidx.room.execSQL
import androidx.room.immediateTransaction
import androidx.room.useWriterConnection
import be.digitalia.fosdem.db.converters.NonNullInstantTypeConverters
import be.digitalia.fosdem.db.entities.EventEntity
import be.digitalia.fosdem.db.entities.EventTitles
import be.digitalia.fosdem.db.entities.EventToPerson
import be.digitalia.fosdem.db.entities.RoomColor
import be.digitalia.fosdem.model.Attachment
import be.digitalia.fosdem.model.Day
import be.digitalia.fosdem.model.DetailedEvent
import be.digitalia.fosdem.model.Event
import be.digitalia.fosdem.model.EventDetails
import be.digitalia.fosdem.model.Link
import be.digitalia.fosdem.model.Person
import be.digitalia.fosdem.model.PersonDetails
import be.digitalia.fosdem.model.ScheduleSection
import be.digitalia.fosdem.model.StatusEvent
import be.digitalia.fosdem.model.Track
import be.digitalia.fosdem.utils.BackgroundWorkScope
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.filterNotNull
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.mapLatest
import kotlinx.coroutines.flow.stateIn
import java.time.Instant

@Dao
abstract class ScheduleDao(private val appDatabase: AppDatabase) {
    val version: Flow<Int> =
        appDatabase.createVersionFlow(BackgroundWorkScope, EventEntity.TABLE_NAME)
    val bookmarksVersion: Flow<Int>
        get() = appDatabase.bookmarksDao.version
    val databaseVersion
        get() = AppDatabase.VERSION

    /**
     * @return The conference id, or null if not available.
     */
    val conferenceId: Flow<String?> = appDatabase.dataStore.data.map { prefs ->
        prefs[CONFERENCE_ID_PREF_KEY]
    }

    /**
     * @return The conference title, or null if not available.
     */
    val conferenceTitle: Flow<String?> = appDatabase.dataStore.data.map { prefs ->
        prefs[CONFERENCE_TITLE_PREF_KEY]
    }

    /**
     * @return The base URL for the schedule website, or null if not available.
     */
    val baseUrl: Flow<String?> = appDatabase.dataStore.data.map { prefs ->
        prefs[BASE_URL_PREF_KEY]
    }

    /**
     * @return The latest update time, or null if not available.
     */
    val latestUpdateTime: Flow<Instant?> = appDatabase.dataStore.data.map { prefs ->
        prefs[LATEST_UPDATE_TIME_PREF_KEY]?.let { Instant.ofEpochMilli(it) }
    }

    /**
     * @return The time identifier of the current version of the database.
     */
    val lastModifiedTag: Flow<String?> = appDatabase.dataStore.data.map { prefs ->
        prefs[LAST_MODIFIED_TAG_PREF_KEY]
    }

    private class EmptyScheduleException : Exception()

    /**
     * Stores the schedule in the database.
     *
     * @param schedule The streaming schedule data split in multiple sections.
     * @return The number of events processed. If 0, no data or metadata will be inserted.
     */
    suspend fun storeSchedule(schedule: Sequence<ScheduleSection>, lastModifiedTag: String?): Int {
        return try {
            appDatabase.useWriterConnection { transactor ->
                transactor.withTransaction(SQLiteTransactionType.EXCLUSIVE) {
                    var totalEvents = 0
                    var minEventId = Long.MAX_VALUE
                    var conferenceSection: ScheduleSection.Conference? = null

                    // 1: Delete the previous schedule
                    clearSchedule()

                    // 2: Store the main schedule data
                    val tracksByName = mutableMapOf<String, Track>()
                    for (section in schedule) {
                        when (section) {
                            is ScheduleSection.Conference -> conferenceSection = section
                            is ScheduleSection.Persons -> storePersonDetails(section.persons)
                            is ScheduleSection.Day -> {
                                val result = storeDay(section.day, section.events, tracksByName)
                                totalEvents += result.totalEvents
                                minEventId = minOf(minEventId, result.minEventId)
                            }
                        }
                    }

                    if (totalEvents == 0) {
                        // Trigger a transaction rollback but don't report an error
                        throw EmptyScheduleException()
                    }

                    // 3: Purge outdated bookmarks and room colors
                    purgeOutdatedBookmarks(minEventId)
                    purgeOutdatedRoomColors()

                    // 4: Store the conference data
                    checkNotNull(conferenceSection) { "Missing conference data" }
                    appDatabase.dataStore.edit { prefs ->
                        prefs.clear()
                        prefs[CONFERENCE_ID_PREF_KEY] = conferenceSection.conferenceId
                        prefs[CONFERENCE_TITLE_PREF_KEY] = conferenceSection.conferenceTitle
                        prefs[BASE_URL_PREF_KEY] = conferenceSection.baseUrl
                        prefs[LATEST_UPDATE_TIME_PREF_KEY] = System.currentTimeMillis()
                        if (lastModifiedTag != null) {
                            prefs[LAST_MODIFIED_TAG_PREF_KEY] = lastModifiedTag
                        }
                    }

                    totalEvents
                }
            }
        } catch (_: EmptyScheduleException) {
            0
        }
    }

    private suspend fun storePersonDetails(persons: Sequence<PersonDetails>) {
        for (person in persons) {
            insertPersonDetails(person)
        }
    }

    private class DayResult(
        val totalEvents: Int,
        val minEventId: Long
    )

    private suspend fun storeDay(
        day: Day,
        dayEvents: Sequence<DetailedEvent>,
        tracksByName: MutableMap<String, Track>
    ): DayResult {
        // 1: Insert the events
        var totalEvents = 0
        var minEventId = Long.MAX_VALUE

        for ((event, details) in dayEvents) {
            // Retrieve or insert Track
            val eventTrack = event.track
            val track = tracksByName.getOrPut(eventTrack.name) {
                // New track
                eventTrack.copy(id = tracksByName.size + 1L).also { insertTrack(it) }
            }

            val eventId = event.id
            try {
                // Insert main event and fulltext fields
                val eventEntity = EventEntity(
                    id = eventId,
                    dayIndex = day.index,
                    startTime = event.startTime,
                    startTimeOffset = event.startTimeOffset,
                    endTime = event.endTime,
                    roomName = event.roomName,
                    url = event.url,
                    trackId = track.id,
                    abstractText = event.abstractText,
                    description = event.description,
                    feedbackUrl = event.feedbackUrl
                )
                val eventTitles = EventTitles(
                    id = eventId,
                    title = event.title,
                    subTitle = event.subTitle
                )
                insertEvent(eventEntity, eventTitles)
            } catch (_: Exception) {
                // Duplicate event: skip
                continue
            }

            if (eventId < minEventId) {
                minEventId = eventId
            }

            val persons = details.persons
            insertPersons(persons)
            val eventsToPersons = persons.map { EventToPerson(eventId, it.id) }
            insertEventsToPersons(eventsToPersons)

            insertAttachments(details.attachments)
            insertLinks(details.links)

            totalEvents++
        }

        // 2: Insert the day if it contains at least one event
        if (totalEvents > 0) {
            insertDay(day)
        }

        return DayResult(
            totalEvents = totalEvents,
            minEventId = minEventId
        )
    }

    @Insert
    protected abstract suspend fun insertTrack(track: Track)

    @Insert
    protected abstract suspend fun insertEvent(eventEntity: EventEntity, eventTitles: EventTitles)

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    protected abstract suspend fun insertPersons(persons: List<Person>)

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    protected abstract suspend fun insertPersonDetails(person: PersonDetails)

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    protected abstract suspend fun insertEventsToPersons(eventsToPersons: List<EventToPerson>)

    @Insert
    protected abstract suspend fun insertAttachments(attachments: List<Attachment>)

    @Insert
    protected abstract suspend fun insertLinks(links: List<Link>)

    @Insert
    protected abstract suspend fun insertDay(days: Day)

    @Query("DELETE FROM bookmarks WHERE event_id < :minEventId")
    protected abstract suspend fun purgeOutdatedBookmarks(minEventId: Long)

    @Query("DELETE FROM room_colors WHERE room_name NOT IN (SELECT DISTINCT room_name FROM events WHERE room_name IS NOT NULL)")
    protected abstract suspend fun purgeOutdatedRoomColors()

    @Query("SELECT * FROM room_colors")
    abstract suspend fun getAllRoomColors(): List<RoomColor>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    abstract suspend fun insertRoomColor(roomColor: RoomColor)

    suspend fun clearSchedule() {
        appDatabase.useWriterConnection { transactor ->
            transactor.immediateTransaction {
                execSQL("DELETE FROM events")
                execSQL("DELETE FROM events_titles")
                execSQL("DELETE FROM persons")
                execSQL("DELETE FROM events_persons")
                execSQL("DELETE FROM attachments")
                execSQL("DELETE FROM links")
                execSQL("DELETE FROM tracks")
                execSQL("DELETE FROM days")
            }
        }
    }

    // Cache days
    @OptIn(ExperimentalCoroutinesApi::class)
    val days: Flow<List<Day>> = appDatabase.invalidationTracker.createFlow(Day.TABLE_NAME)
        .mapLatest { getDaysInternal() }
        .stateIn(
            scope = BackgroundWorkScope,
            started = SharingStarted.Lazily,
            initialValue = null
        ).filterNotNull()

    @Query("SELECT `index`, date, start_time, end_time FROM days ORDER BY `index` ASC")
    protected abstract suspend fun getDaysInternal(): List<Day>

    suspend fun getYear(): Int? {
        // Compute from days if available
        return days.first().firstOrNull()?.date?.year
    }

    @Query("""SELECT t.id, t.name, t.type FROM tracks t
        JOIN events e ON t.id = e.track_id
        WHERE e.day_index = :day
        GROUP BY t.id
        ORDER BY t.name ASC""")
    abstract suspend fun getTracks(day: Day): List<Track>

    /**
     * Returns the event with the specified id, or null if not found.
     */
    @Query("SELECT * FROM events_view WHERE id = :id")
    abstract suspend fun getEvent(id: Long): Event?

    /**
     * Returns all found events whose id is part of the given list.
     */
    @Query("""SELECT ev.*, b.event_id IS NOT NULL AS is_bookmarked
        FROM events_view ev
        LEFT JOIN bookmarks b ON ev.id = b.event_id
        WHERE ev.id IN (:ids)
        ORDER BY ev.start_time ASC""")
    abstract fun getEvents(ids: LongArray): PagingSource<Int, StatusEvent>

    /**
     * Returns the events for a specified track, including their bookmark status.
     */
    @Query("""SELECT ev.*, b.event_id IS NOT NULL AS is_bookmarked
        FROM events_view ev
        LEFT JOIN bookmarks b ON ev.id = b.event_id
        WHERE ev.day_index = :day AND ev.track_id = :track
        ORDER BY ev.start_time ASC""")
    abstract suspend fun getEvents(day: Day, track: Track): List<StatusEvent>

    /**
     * Returns the events for a specified track, without their bookmark status.
     */
    @Query("SELECT * FROM events_view WHERE day_index = :day AND track_id = :track ORDER BY start_time ASC")
    abstract suspend fun getEventsWithoutBookmarkStatus(day: Day, track: Track): List<Event>

    /**
     * Returns events starting in the specified interval, ordered by ascending start time.
     */
    @Query("""SELECT ev.*, b.event_id IS NOT NULL AS is_bookmarked
        FROM events_view ev
        LEFT JOIN bookmarks b ON ev.id = b.event_id
        WHERE ev.start_time BETWEEN :minStartTime AND :maxStartTime
        ORDER BY ev.start_time ASC""")
    @TypeConverters(NonNullInstantTypeConverters::class)
    abstract fun getEventsWithStartTime(minStartTime: Instant, maxStartTime: Instant): PagingSource<Int, StatusEvent>

    /**
     * Returns events in progress at the specified time, ordered by descending start time.
     */
    @Query("""SELECT ev.*, b.event_id IS NOT NULL AS is_bookmarked
        FROM events_view ev
        LEFT JOIN bookmarks b ON ev.id = b.event_id
        WHERE ev.start_time <= :time AND :time < ev.end_time
        ORDER BY ev.start_time DESC""")
    @TypeConverters(NonNullInstantTypeConverters::class)
    abstract fun getEventsInProgress(time: Instant): PagingSource<Int, StatusEvent>

    /**
     * Returns the events presented by the specified person.
     */
    @Query("""SELECT ev.*, b.event_id IS NOT NULL AS is_bookmarked
        FROM events_view ev
        LEFT JOIN bookmarks b ON ev.id = b.event_id
        JOIN events_persons ep2 ON ev.id = ep2.event_id
        WHERE ep2.person_id = :person
        ORDER BY ev.start_time ASC""")
    abstract fun getEvents(person: Person): PagingSource<Int, StatusEvent>

    /**
     * Search through matching titles, subtitles, track names, person names.
     * We need to use an union of 3 sub-queries because a "match" condition can not be
     * accompanied by other conditions in a "where" statement.
     */
    @Query("""SELECT ev.*, b.event_id IS NOT NULL AS is_bookmarked
        FROM events_view ev
        LEFT JOIN bookmarks b ON ev.id = b.event_id
        WHERE ev.id IN (
            SELECT `rowid`
            FROM events_titles
            WHERE events_titles MATCH :query || '*'
        UNION
            SELECT e.id
            FROM events e
            JOIN tracks t ON e.track_id = t.id
            WHERE t.name LIKE '%' || :query || '%'
        UNION
            SELECT ep.event_id
            FROM events_persons ep
            JOIN persons p ON ep.person_id = p.`rowid`
            WHERE p.name MATCH :query || '*'
        )
        ORDER BY ev.start_time ASC""")
    abstract fun getSearchResults(query: String): PagingSource<Int, StatusEvent>

    /**
     * Returns all persons in alphabetical order.
     */
    @Query("""SELECT `rowid`, name
        FROM persons
        ORDER BY name COLLATE NOCASE""")
    abstract fun getPersons(): PagingSource<Int, Person>

    /**
     * Returns the details of the given person, if available.
     */
    @Query("SELECT * FROM person_details WHERE `rowid` = :person")
    abstract suspend fun getPersonDetails(person: Person): PersonDetails?

    suspend fun getEventDetails(event: Event): EventDetails {
        // Load persons, attachments and links in parallel
        return coroutineScope {
            val persons = async { getPersons(event) }
            val attachments = async { getAttachments(event) }
            val links = async { getLinks(event) }
            EventDetails(
                persons = persons.await(),
                attachments = attachments.await(),
                links = links.await()
            )
        }
    }

    @Query("""SELECT p.`rowid`, p.name
        FROM persons p
        JOIN events_persons ep ON p.`rowid` = ep.person_id
        WHERE ep.event_id = :event""")
    protected abstract suspend fun getPersons(event: Event): List<Person>

    @Query("SELECT * FROM attachments WHERE event_id = :event ORDER BY id ASC")
    protected abstract suspend fun getAttachments(event: Event?): List<Attachment>

    @Query("SELECT * FROM links WHERE event_id = :event ORDER BY id ASC")
    protected abstract suspend fun getLinks(event: Event?): List<Link>

    companion object {
        private val CONFERENCE_ID_PREF_KEY = stringPreferencesKey("conference_id")
        private val CONFERENCE_TITLE_PREF_KEY = stringPreferencesKey("conference_title")
        private val BASE_URL_PREF_KEY = stringPreferencesKey("base_url")
        private val LATEST_UPDATE_TIME_PREF_KEY = longPreferencesKey("latest_update_time")
        private val LAST_MODIFIED_TAG_PREF_KEY = stringPreferencesKey("last_modified_tag")
    }
}